test/conftest.py

```diff
@@ -56,3 +58,17 @@ def non_deterministic_case(
 @fixture(params=ADJOINT_CASES)
 def adjoint(request) -> bool:
     return request.param
+
+
+@fixture(params=KFAC_EXPAND_EXACT_CASES)
+def kfac_expand_exact_case(
```

Comment:
I don't fully understand why this function is necessary, could you briefly explain?

Reply:
Functions in `conftest.py` are fixtures, i.e. they can be used by test functions. This happens below where `kfac_expand_exact_case` is used as argument of a test function. The test will then iterate through all cases yielded by `kfac_expand_exact_case`. This is similar to `mark.parametrize`. One advantage is that we can also do things like fixing the random seed in the fixture, so the setup code does not leak into the test function.

Reply:
Disclaimer: I experimented with `fixture`s in this project. Not sure if they are the best way though... If you believe this is too complicated, I'm okay adding new tests in a `fixture`-free way.

Reply:
I see, thanks! I'm fine with this design, not sure if the advantages are worth the additional complexity, but I'm happy to try it for this project. We can still simplify later.